pages/posts/2017/12-december/140-continuous-integration-at-subvisual.md

```diff
@@ -0,0 +1,116 @@
+
+Nowadays, we dropped the more complex approaches in favor of a simpler one. We have a `master` branch, from which all other branches are created, and to which they are all merged to. That's it.
+
+Branches usually have an owner and are only touched by that owner. But we also have no problem sharing them if it suits the needs. For example, I may start the back-end logic required for a new feature, and then hand over the branch so that a front-end developer and a designer can work on the rest. Or we might pair program on it if we feel the need for more coordination.
+
+Once we merge our changes, Semaphore will take care of checking the new build and, if automatic deploys are enabled, put it live.
+
+Of course, "with great power, comes great responsibility". We don't just merge to `master` whenever we wish to. But instead of a more rigorous git workflow, we rely on a strong automated-testing culture. GitHub and Semaphore also help a great deal with the process as well.
+
+# Code Reviews
+
+We've been doing code reviews and enforcing them as required for all of our projects for a long time now. Our rule is simple: every change that is merged must be approved by at least one person other than yourself. So you either need someone else to review it, or you can pair with someone during development.
+
+And with our tendency for noticing every single detail, it's rare for a single extra newline to go unnoticed. 
+
+![2017-10-29-181705_790x539_scrot.png](https://s3.eu-west-1.amazonaws.com/subvisual/blog/posts/140/github1.png)
+We're extremely paranoid about the quality of our code.
+
+Recent updates to Github helped this a lot, with Code Reviews now being a part of the Pull Requests interface.
+
+![2017-10-29-181817_783x430_scrot.png](https://s3.eu-west-1.amazonaws.com/subvisual/blog/posts/140/github2.png)
+
+# CI builds
+
+One of the reasons we've chosen Semaphore when we first got started was their great ability to customize the setup and build process. We can create our own bash script to customize the build, and even install additional packages if we need to (especially useful if we're using edge versions of a library, and need to compile them from source). Here's an example for a heavily-customized Elixir/React application:
+
+```bash
+# setup
+source $HOME/.kerl/installs/19/activate
+nvm use $NODE_VERSION
+mix local.hex --force
+mix local.rebar --force
+mix deps.get
+yarn install
+MIX_ENV=test mix ecto.setup
+
+# build
+mix credo
+mix test --exclude feature
+mix test --only feature # trust us. there's a reason for this
+yarn lint
+```
+
+It also allows us to customize what directories get cached between builds.
+
+These two features combined make it a great resource even if your framework or language isn't yet officially supported by them. You can input the commands to download and compile such dependencies, and then cache them so that it doesn't slow down your future builds.
```

Comment:
The last time you referred Semaphore was 2 paragraphs ago, in the beginning, so by now I feel like the *it* and *them* have become fuzzy (especially if you've gone through the trouble of reading the script), so I suggest changing to `... combined make Semaphore a great resource...`.

Reply:
fixed